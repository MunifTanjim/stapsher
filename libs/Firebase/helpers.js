const _ = require('lodash')

const { store, fieldValue } = require('../Firebase')

const { normalizeRepos } = _require('libs/GitHub/helpers/payload')

const logger = _require('libs/Logger')

const createInstallationOnStore = async (installation, repos) => {
  let repoIDs = _.map(repos, 'id')
  let flattenedRepos = _.keyBy(normalizeRepos(repos, installation.id), 'id')

  try {
    let batch = store.batch()

    batch.set(store.collection('installations').doc(String(installation.id)), {
      installation,
      repositories: _.keyBy(repos, 'id')
    })

    let reposCollection = store.collection('repositories')

    repoIDs.forEach(repoID => {
      batch.set(reposCollection.doc(String(repoID)), flattenedRepos[repoID])
    })

    return await batch.commit()
  } catch (err) {
    throw err
  }
}

const deleteInstallationFromStore = async (installation, repos) => {
  try {
    let batch = store.batch()

    batch.delete(store.collection('installations').doc(String(installation.id)))

    let reposSnapshot = await store
      .collection('repositories')
      .where('installation_id', '==', installation.id)
      .get()

    reposSnapshot.forEach(repoDoc => {
      batch.delete(repoDoc.ref)
    })

    return await batch.commit()
  } catch (err) {
    throw err
  }
}

const addReposToStore = async (installation, repos) => {
  let repoIDs = _.map(repos, 'id')
  let flattenedRepos = _.keyBy(normalizeRepos(repos, installation.id), 'id')

  try {
    let batch = store.batch()

    batch.set(
      store.collection('installations').doc(String(installation.id)),
      {
        repositories: _.keyBy(repos, 'id')
      },
      { merge: true }
    )

    let repositoriesCollection = store.collection('repositories')

    repoIDs.forEach(repoID => {
      batch.set(
        repositoriesCollection.doc(String(repoID)),
        flattenedRepos[repoID]
      )
    })

    return await batch.commit()
  } catch (err) {
    throw err
  }
}

const removeReposFromStore = async (installation, repos) => {
  let repoIDs = _.map(repos, 'id')

  try {
    let batch = store.batch()

    batch.update(
      store.collection('installations').doc(String(installation.id)),
      _.assign(
        {},
        ...repoIDs.map(repoID => ({
          [`repositories.${repoID}`]: fieldValue.delete()
        }))
      )
    )

    let repositoriesCollection = store.collection('repositories')

    repoIDs.forEach(repoID => {
      batch.delete(repositoriesCollection.doc(String(repoID)))
    })

    return await batch.commit()
  } catch (err) {
    throw err
  }
}

const fetchInstallationIdFromStore = async ({ username, repository }) => {
  logger.verbose('Fetching installation_id from FireStore')

  try {
    let { docs } = await store
      .collection('repositories')
      .where('full_name', '==', `${username}/${repository}`)
      .get()

    let id = docs.length ? docs[0].data().installation_id : null

    return id
  } catch (err) {
    throw err
  }
}

module.exports = {
  createInstallationOnStore,
  deleteInstallationFromStore,
  addReposToStore,
  removeReposFromStore,
  fetchInstallationIdFromStore
}
