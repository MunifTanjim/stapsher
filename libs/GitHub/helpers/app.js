const isNull = require('lodash.isnull')
const parseLinkHeader = require('parse-link-header')

const cache = _require('libs/lowdb')
const logger = _require('libs/Logger')
const { ResponseError } = _require('libs/Error')
const { fetchInstallationIdFromStore } = _require('libs/Firebase/helpers')
const { fetchInstallationIdFromCache } = _require('libs/lowdb/helpers')

const fetchInstallationIdFromGitHub = async ({ username }, authAsGithubApp) => {
  logger.verbose('Getting installation_id from GitHub API')

  try {
    let api = await authAsGithubApp()

    let page = 1,
      hasNext = true

    while (hasNext) {
      let { data, meta } = await api.apps.getInstallations({
        page
      })

      for ({ id, account } of data) {
        if (account.login === username) return id
      }

      if (meta.link) {
        let { next } = parseLinkHeader(meta.link)
        if (next) page = next.page
        else hasNext = false
      } else hasNext = false
    }

    return null
  } catch (err) {
    throw err
  }
}

const fetchInstallationId = async (info, authAsGithubApp) => {
  logger.verbose('Getting installation_id')

  let id

  try {
    id = await fetchInstallationIdFromCache(info)
    if (!isNull(id)) return id

    id = await fetchInstallationIdFromStore(info)
    if (!isNull(id)) return id

    id = await fetchInstallationIdFromGitHub(info, authAsGithubApp)
    if (!isNull(id)) return id

    throw new ResponseError('GITHUB_APP_NOT_INSTALLED', 400)
  } catch (err) {
    throw err
  }
}

module.exports = {
  fetchInstallationId
}
