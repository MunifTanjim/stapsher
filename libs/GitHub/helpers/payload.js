const normalizeRepos = (repos, installation_id) =>
  repos.map(repo => ({
    ...repo,
    installation_id
  }))

const getInstallation = ({ installation: { id, account } }) => ({
  id,
  account: {
    id: account.id,
    login: account.login,
    type: account.type
  }
})

const getRepositories = payload => {
  let { action } = payload
  switch (action) {
    case 'created':
      return payload[`repositories`]
    case 'deleted':
      return []
    case 'added':
    case 'removed':
      return payload[`repositories_${action}`]
  }
}

module.exports = {
  normalizeRepos,
  getInstallation,
  getRepositories
}
