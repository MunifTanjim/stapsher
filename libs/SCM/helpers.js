const yaml = require('js-yaml')

const { throwError } = require('../Error')

const parseFile = (blob, extension) => {
  try {
    switch (extension.toLowerCase()) {
      case 'json':
        return JSON.parse(blob)
      case 'yaml':
      case 'yml':
        return yaml.safeLoad(blob, 'utf8')
      default:
        throwError('UNSUPPORTED_EXTENSION', { extension }, 422)
    }
  } catch (err) {
    if (err instanceof SyntaxError) {
      throwError('FILE_PARSE_FAILED', err, 422)
    } else {
      throw err
    }
  }
}

module.exports.parseFile = parseFile
