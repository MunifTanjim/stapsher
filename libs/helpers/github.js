const { store } = require('../Firebase')
const cache = require('../lowdb')

const parseLinkHeader = require('parse-link-header')

const logger = require('../Logger')

const normalizeRepos = (repos, installation_id) =>
  repos.map(repo => ({
    ...repo,
    installation_id
  }))

const getInstallation = ({ installation: { id, account } }) => ({
  id,
  account: {
    id: account.id,
    login: account.login,
    type: account.type
  }
})

const getRepositories = payload => {
  let { action } = payload
  switch (action) {
    case 'created':
      return payload[`repositories`]
    case 'deleted':
      return []
    case 'added':
    case 'removed':
      return payload[`repositories_${action}`]
  }
}

const fetchInstallationIdFromStore = async ({ username, repository }) => {
  logger.verbose('Getting installation_id from FireStore')

  try {
    let { docs } = await store
      .collection('repositories')
      .where('full_name', '==', `${username}/${repository}`)
      .get()

    let id = docs.length ? docs[0].data().installation_id : null

    return Promise.resolve(id)
  } catch (err) {}
}
const fetchInstallationIdFromCache = async ({ username, repository }) => {
  logger.verbose('Getting installation_id from lowdb Cache')

  try {
    let db = await cache()

    let id =
      db
        .get('repositories')
        .find({ full_name: `${username}/${repository}` })
        .get('installation_id')
        .value() || null

    return Promise.resolve(id)
  } catch (err) {}
}

const fetchInstallationIdFromGitHub = async ({ username }, api) => {
  logger.verbose('Getting installation_id from GitHub API')

  try {
    let page = 1,
      hasNext = true

    while (hasNext) {
      let { data, meta } = await api.apps.getInstallations({
        page
      })

      for ({ id, account } of data) {
        if (account.login === username) return Promise.resolve(id)
      }

      if (meta.link) {
        let { next } = parseLinkHeader(meta.link)
        if (next) page = next.page
        else hasNext = false
      } else hasNext = false
    }

    return Promise.resolve(null)
  } catch (err) {}
}

const fetchInstallationId = async (info, api) => {
  logger.verbose('Getting installation_id')

  let id

  try {
    id = await fetchInstallationIdFromCache(info)
    if (Boolean(id)) return Promise.resolve(id)

    id = await fetchInstallationIdFromStore(info)
    if (Boolean(id)) return Promise.resolve(id)

    id = await fetchInstallationIdFromGitHub(info, api)
    if (Boolean(id)) return Promise.resolve(id)

    return Promise.resolve(null)
  } catch (err) {
    return Promise.reject(err)
  }
}

module.exports.payload = {
  normalizeRepos,
  getInstallation,
  getRepositories
}

module.exports.app = {
  fetchInstallationId
}
