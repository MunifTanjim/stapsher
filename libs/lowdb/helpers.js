const _ = require('lodash')

const cache = require('../lowdb')

const { normalizeRepos } = _require('libs/GitHub/helpers/payload')

const logger = _require('libs/Logger')

const createInstallationOnCache = async (installation, repos) => {
  let flattenedRepos = _.keyBy(normalizeRepos(repos, installation.id), 'id')

  try {
    let db = await cache()

    db
      .set(`installations[${installation.id}]`, {
        installation,
        repositories: _.keyBy(repos, 'id')
      })
      .write()

    db
      .get('repositories')
      .extend(flattenedRepos)
      .write()

    return true
  } catch (err) {
    throw err
  }
}

const deleteInstallationFromCache = async (installation, repos) => {
  try {
    let db = await cache()

    let repoIds = db
      .get(`installations[${installation.id}].repositories`)
      .keys()
      .value()

    repoIds.forEach(repoId => {
      db
        .get('repositories')
        .unset(repoId)
        .write()
    })

    db
      .get(`installations`)
      .unset(installation.id)
      .write()

    return true
  } catch (err) {
    throw err
  }
}

const addReposToCache = async (installation, repos) => {
  try {
    let db = await cache()

    db
      .get(`installations[${installation.id}].repositories`)
      .extend(_.keyBy(repos, 'id'))
      .write()

    db
      .get('repositories')
      .extend(_.keyBy(normalizeRepos(repos, installation.id), 'id'))
      .write()

    return true
  } catch (err) {
    throw err
  }
}

const removeReposFromCache = async (installation, repos) => {
  let repoIDs = _.map(repos, 'id')

  try {
    let db = await cache()

    repoIDs.forEach(repoID => {
      db
        .get(`installations[${installation.id}].repositories`)
        .unset(repoID)
        .write()

      db
        .get('repositories')
        .unset(repoID)
        .write()
    })

    return true
  } catch (err) {
    throw err
  }
}

const fetchInstallationIdFromCache = async ({ username, repository }) => {
  logger.verbose('Fetching installation_id from lowdb Cache')

  try {
    let db = await cache()

    let id =
      db
        .get('repositories')
        .find({ full_name: `${username}/${repository}` })
        .get('installation_id')
        .value() || null

    return id
  } catch (err) {
    throw err
  }
}

module.exports = {
  createInstallationOnCache,
  deleteInstallationFromCache,
  addReposToCache,
  removeReposFromCache,
  fetchInstallationIdFromCache
}
