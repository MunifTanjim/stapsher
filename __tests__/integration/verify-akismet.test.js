const fetch = require('node-fetch')

require('./helpers/init')
const helpers = require('../helpers')

jest.mock('../../libs/Stapsher')
const Stapsher = require('../../libs/Stapsher')

const akismet = require('../../libs/Akismet')
akismet.akismetVerify = jest.fn()

const app = require('../../app')

let server
beforeAll(done => {
  server = helpers.startServer(done, app)
})
afterAll(done => {
  helpers.stopServer(done, server)
})

beforeEach(() => {
  Stapsher.mockClear()
})

describe('Stapsher:integration:verify-akismet', () => {
  let baseUrl = helpers.getBaseUrl()
  let parameters = helpers.getParameters()

  let { platform, username, repository, branch, entryType } = parameters
  let url = `${baseUrl}/${platform}/${username}/${repository}/${branch}/${entryType}/verify/akismet`

  let mockConfig = {
    'akismet.apiKey': 'qwerty',
    'akismet.siteUrl': 'http://example.com'
  }

  Stapsher.prototype.getConfig.mockImplementation(() => ({
    get: key => mockConfig[key]
  }))

  it('responses when disabled', async () => {
    mockConfig['akismet.enable'] = false

    let res = await fetch(url)

    expect(Stapsher).toHaveBeenCalledTimes(1)
    expect(Stapsher.mock.instances[0].authenticate).toHaveBeenCalledTimes(1)
    expect(Stapsher.mock.instances[0].getConfig).toHaveBeenCalledTimes(1)
    expect(await res.json()).toMatchSnapshot()
  })

  it('responses when valid', async () => {
    akismet.akismetVerify.mockResolvedValueOnce(true)

    mockConfig['akismet.enable'] = true

    let res = await fetch(url)

    expect(Stapsher).toHaveBeenCalledTimes(1)
    expect(Stapsher.mock.instances[0].authenticate).toHaveBeenCalledTimes(1)
    expect(Stapsher.mock.instances[0].getConfig).toHaveBeenCalledTimes(1)
    expect(akismet.akismetVerify).toHaveBeenCalledTimes(1)
    expect(akismet.akismetVerify).toHaveBeenLastCalledWith(
      mockConfig['akismet.apiKey'],
      mockConfig['akismet.siteUrl']
    )
    expect(await res.json()).toMatchSnapshot()
  })

  it('responses when invalid', async () => {
    akismet.akismetVerify.mockResolvedValueOnce(false)

    mockConfig['akismet.enable'] = true

    let res = await fetch(url)

    expect(Stapsher).toHaveBeenCalledTimes(1)
    expect(Stapsher.mock.instances[0].authenticate).toHaveBeenCalledTimes(1)
    expect(Stapsher.mock.instances[0].getConfig).toHaveBeenCalledTimes(1)
    expect(akismet.akismetVerify).toHaveBeenCalledTimes(2)
    expect(akismet.akismetVerify).toHaveBeenLastCalledWith(
      mockConfig['akismet.apiKey'],
      mockConfig['akismet.siteUrl']
    )
    expect(await res.json()).toMatchSnapshot()
  })
})
