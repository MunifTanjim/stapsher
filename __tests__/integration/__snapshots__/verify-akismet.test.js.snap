// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stapsher:integration:verify-akismet responds when disabled 1`] = `
{
  "enabled": false
}
`;

exports[`Stapsher:integration:verify-akismet responds when invalid 1`] = `
{
  "enabled": true,
  "validAPIKey": false
}
`;

exports[`Stapsher:integration:verify-akismet responds when valid 1`] = `
{
  "enabled": true,
  "validAPIKey": true
}
`;
